grpc/shared_conn.go

```diff
@@ -18,68 +24,190 @@ type OnTrackCB func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver)
-	sc := &SharedConn{
-		ReconfigurableClientConn: conn,
-		peerConn:                 peerConn,
-		resourceOnTrackCBs:       map[resource.Name]OnTrackCB{},
-		logger:                   logger,
-	}
-	if peerConn != nil {
-		peerConn.OnTrack(func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver) {
-			name, err := resource.NewFromString(tr.StreamID())
-			if err != nil {
-				logger.Errorf("%p OnTrack called with StreamID: %s which is not able to be parsed into a resource name", sc, tr.StreamID())
-				return
-			}
-			sc.mu.RLock()
-			onTrackCB, ok := sc.resourceOnTrackCBs[name]
-			sc.mu.RUnlock()
-			if !ok {
-				sc.logger.Errorf("%p OnTrack called with StreamID: %s which is not in the resourceOnTrackCBs: %#v", sc, name, sc.resourceOnTrackCBs)
-				return
-			}
-			onTrackCB(tr, r)
-		})
-	}
-	return sc
+// Invoke forwards to the underlying GRPC Connection.
+func (sc *SharedConn) Invoke(
+	ctx context.Context,
+	method string,
+	args, reply interface{},
+	opts ...googlegrpc.CallOption,
+) error {
+	return sc.grpcConn.Invoke(ctx, method, args, reply, opts...)
+}
+
+// NewStream forwards to the underlying GRPC Connection.
+func (sc *SharedConn) NewStream(
+	ctx context.Context,
+	desc *googlegrpc.StreamDesc,
+	method string,
+	opts ...googlegrpc.CallOption,
+) (googlegrpc.ClientStream, error) {
+	return sc.grpcConn.NewStream(ctx, desc, method, opts...)
 }
 
 // AddOnTrackSub adds an OnTrack subscription for the resource.
 func (sc *SharedConn) AddOnTrackSub(name resource.Name, onTrackCB OnTrackCB) {
-	sc.mu.Lock()
-	defer sc.mu.Unlock()
-	sc.resourceOnTrackCBs[name] = onTrackCB
+	sc.resOnTrackMu.Lock()
+	defer sc.resOnTrackMu.Unlock()
+	sc.resOnTrackCBs[name] = onTrackCB
 }
 
 // RemoveOnTrackSub removes an OnTrack subscription for the resource.
 func (sc *SharedConn) RemoveOnTrackSub(name resource.Name) {
-	sc.mu.Lock()
-	defer sc.mu.Unlock()
-	delete(sc.resourceOnTrackCBs, name)
+	sc.resOnTrackMu.Lock()
+	defer sc.resOnTrackMu.Unlock()
+	delete(sc.resOnTrackCBs, name)
+}
+
+// GrpcConn returns a gRPC capable client connection.
+func (sc *SharedConn) GrpcConn() *ReconfigurableClientConn {
+	return &sc.grpcConn
 }
 
-// PeerConn returns the PeerConnection.
+// PeerConn returns a WebRTC PeerConnection capable of sending video/audio data.
 func (sc *SharedConn) PeerConn() *webrtc.PeerConnection {
 	return sc.peerConn
 }
 
+// ResetConn replaces the underlying GrpcConnection object. It also re-initiatlizes the
+// PeerConnection that must be renegotiated.
+func (sc *SharedConn) ResetConn(conn rpc.ClientConn, moduleLogger logging.Logger) {
+	sc.grpcConn.ReplaceConn(conn)
+
+	sc.logger = moduleLogger.Sublogger("conn")
+
+	peerConn, err := webrtc.NewPeerConnection(webrtc.Configuration{})
+	if err != nil {
+		sc.logger.Warnw("Unable to create optional peer connection for module. Ignoring.", "err", err)
+		return
+	}
+
+	sc.peerConn = peerConn
```

Comment:
if `sc.peerConn` was not nil before, won't just replacing it like this without closing it or anything leave it in a zombie state?

Reply:
The documented contract for `SharedConn` is to alternate calls between resetconn -> close. That is what I've also observed, but it'd hard to be 100% sure on this.

This is a case where I'd like to have an invariant for testing to ensure we don't hit the state of `peerConn` being active and `ResetConn` being called.

That said, I'm not sure if there's practical harm in either:
* Just closing it if we see it open
* Letting it leak